### ISSUE TITLE
Incorrect checking of generic local functions

### ISSUE THREAD

**Issue by kmillikin** [microsoft/pyright#4285]
In this example:

```python
from typing import Callable, Optional, TypeVar

T = TypeVar('T')

def make_id() -> Callable[[T], T]:
    x: Optional[T] = None
    def not_id(y: T) -> T:
        nonlocal x
        if x is None:
            x = y
        return x
    return not_id

id = make_id()

reveal_type(id('hest'))
reveal_type(id(3.14159))
```
It will reveal `str`, `float` but the code will actually return `str` both times.  It looks like the return type annotation is causing `T` to appear bound in the body but that isn't right.  It should occur free both in `x: Optional[T]` and the signature `(y: T) -> T`.

The expected behavior is similar to the corresponding example with the initialization of `x` lifted out and `nonlocal` changed to `global`, which does report `error: Type variable "T" has no meaning in this context (reportGeneralTypeIssues)`.

**Comment by erictraut** [comment]
This code sample demonstrates an area of ambiguity in the Python typing spec. PEP 484 discusses "scoping" of type variables. A type variable can be scoped to a function (i.e. a callable), a class, or a type alias. In the case of `make_id`, there are two potential interpretations:
1. `T` should be scoped to `make_id`
2. `T` should be scoped to the `Callable` returned by `make_id`

There's no way to distinguish between these two cases in the type system today. There has been some discussion in the typing forums on this topic, but there hasn't been any resolution.

Because this isn't spec'ed, different type checkers handle this subtly differently. Pyright's current approach is to bind `T` to `make_id` and then "detach" it (and rebind it) to the `Callable`. This allows `T` to be solved independently when the resulting `Callable` is called.

But your sample demonstrates a typing hole in this approach. Mypy appears to use a different (better) approach here. It scopes `T` to the `Callable` but not to `make_id` and generates errors when you attempt to reference `T` within the `make_id` body.

I'll need to think about this more, but my inclination is to mirror mypy's behavior here, in which case the code above would generate errors in the `make_id` body. I'll need to think about the implications for PEP 695.

**Comment by kmillikin** [comment]
The first interpretation, that `make_id` is a generic function, isn't very useful, because generic functions are implicitly instantiated at their call sites.  Since the type variable only occurs in the return type hint, there are no constraints at all on the type variable.  And because it occurs both co- and contravariantly, there is no least or greatest choice for `T` with respect to subtyping, so it's not clear how we should choose a solution.

So I've come around to the second interpretation, though I'm pretty sure that's not what PEP-484 had originally intended.

If we had explicit syntax, we might say that the alternatives are:

```python
def make_id[T]() -> Callable[[T], T]:  # make_id is generic in T, returns a non-generic callable
```
and
```python
def make_id() -> lambda T: Callable[[T], T]: # make_id returns a generic callable type
```
And the second one is the interpretation we seem to have settled on as a default.  In that case, the `lambda` makes clear that the `T` ought to be bound in the return type annotation but not in the body of `make_id`.

There probably needs to be an internal representation of a generic function type.  The names of the bound variables don't matter and a generic function type is a subtype of all of its instantiations.  Internally distinguishing where the variables are bound:

```python
def make_id[T]() -> Callable[[T], T]:  # 1
  def id(x: T) -> T: return x
  return id

def make_id() -> lambda T: Callable[[T], T]:  #2
  def id(x: S) -> S: return x
  return id

def make_id[T]() -> Callable[[T], T]:  #3
  def id(x: S) -> S: return x
  return id

def mak_id() -> lambda T: Callable[[T], T]: #4
  def id(x: T) -> T: return x
  return id
```

#1 `make_id` is a generic function that returns a non-generic callable.  The local function `id` is not generic, `T` will be bound by an instantiation of `make_id`.

#2 is a non-generic function that return a generic callable.  The local function id is generic and the name of the bound type variable doesn't matter.  `lambda S: Callable[[S], S]` is the same as `lambda T: Callable[[T], T]`.

#3 is a generic function returning a non-generic callable.  The local function `id` is generic with type `lambda S: Callable[[S], S]` but that's a subtype of all its instantiations, namely `Callable[[T], T]` for any `T`.

#4 is a non-generic function that returns a generic callable.  It's exactly the same as #2 and the bound variables happen to have the same name.

**Comment by erictraut** [comment]
This will be addressed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.286, which I just published. It will also be included in a future version of pylance.

**Comment by erictraut** [comment]
In retrospect, this change was a bad idea, and I've decided to reverse this change. This is partly because of the ongoing efforts to clarify and increase conformity with the typing spec. It also makes sense for consistency with PEP 695. And the behavior associated with this change results in various problems such as that described in #7722.

Going forward, pyright will strictly follow the TypeVar scoping rules as specified in the typing spec and will scope TypeVars to a function that uses those TypeVars in its signature even if they are found only within the return type annotation. Any reference to that TypeVar within an inner scope refers to the outer-scoped TypeVar.

The underlying issue here is that there's a missing feature in the type system. It's currently not possible to specify that a TypeVar should be scoped only to a `Callable`. There are [ongoing discussions](https://discuss.python.org/t/reviving-the-hybrid-keyword-arrow-syntax-for-callables/52499/9) about potential ways to address this shortcoming. However, it will require a new PEP and an update to the typing spec.
